File: systems-for-design-and-implementation/labs/lab3/LibraryPersistance/src/library/persistance/repository/jdbc/UserRepositoryJDBC.scala
package library.persistance.repository.jdbc

import java.sql.{Connection, PreparedStatement, ResultSet, SQLException}
import java.util

import library.model.{User, UserException}
import library.persistance.repository.{IUserRepository, RepositoryException}

class UserRepositoryJDBC extends IUserRepository
{
    def verifyUser(user: User): Boolean =
    {
        val connection: Connection = DBConnection.get.getConnection
        assert(connection != null)
        try
        {
            val stmt: PreparedStatement = connection
                .prepareStatement("SELECT * FROM users WHERE username=? AND password=?")
            stmt.setString(1, user.getUsername)
            stmt.setString(2, user.getPassword)
            val rs: ResultSet = stmt.executeQuery
            rs.next
        }
        catch
        {
            case e: SQLException =>
                throw new UserException("Error " + e)
        }
    }

    def getUserByUsername(username: String): User =
    {
        val connection: Connection = DBConnection.get.getConnection
        var user: User = null
        try
        {
            val stmt: PreparedStatement = connection.prepareStatement("SELECT * FROM users WHERE username = ?")
            stmt.setString(1, username)
            val rs: ResultSet = stmt.executeQuery
            if (rs.next)
            {
                val id: Int = rs.getInt("id")
                val is_librarian: Boolean = rs.getInt("is_librarian") == 1
                user = new User(id, rs.getString("username"), rs.getString("password"), is_librarian,
                    getBorrowedBooks(id))
            }
            stmt.close()
        }
        catch
        {
            case e: SQLException => throw new RepositoryException("Error " + e)
        }
        user
    }

    private def getBorrowedBooks(user_id: Int): util.Set[Integer] =
    {
        val connection: Connection = DBConnection.get.getConnection
        val books: util.Set[Integer] = new util.HashSet[Integer]
        try
        {
            val stmt: PreparedStatement = connection.prepareStatement("SELECT book_id FROM loans WHERE user_id = ?")
            stmt.setInt(1, user_id)
            val rs: ResultSet = stmt.executeQuery
            while (rs.next)
            {
                {
                    books.add(rs.getInt("book_id"))
                }
            }
            stmt.close()
        }
        catch
        {
            case e: SQLException => throw new RepositoryException("Error " + e)
        }
        books
    }
}

File: systems-for-design-and-implementation/labs/lab2/LibraryPersistance/src/library/persistance/repository/jdbc/UserRepositoryJDBC.java
package library.persistance.repository.jdbc;

import library.model.User;
import library.model.UserException;
import library.persistance.repository.IUserRepository;
import library.persistance.repository.RepositoryException;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashSet;
import java.util.Set;

public class UserRepositoryJDBC implements IUserRepository
{
    private Set<Integer> getBorrowedBooks(int user_id)
    {
        Connection connection = DBConnection.get().getConnection();
        Set<Integer> books = new HashSet<Integer>();

        try
        {
            PreparedStatement stmt = connection.prepareStatement("SELECT book_id FROM loans WHERE user_id = ?");
            stmt.setInt(1, user_id);
            ResultSet rs = stmt.executeQuery();
            while (rs.next())
            {
                books.add(rs.getInt("book_id"));
            }
            stmt.close();
        }
        catch (SQLException e)
        {
            throw new RepositoryException("Error " + e);
        }

        return books;
    }

    @Override
    public boolean verifyUser(User user)
    {
        Connection connection = DBConnection.get().getConnection();
        try
        {
            PreparedStatement stmt = connection.prepareStatement("SELECT * FROM users WHERE username=? AND password=?");
            stmt.setString(1, user.getUsername());
            stmt.setString(2, user.getPassword());

            ResultSet rs = stmt.executeQuery();
            return rs.next();
        }
        catch (SQLException e)
        {
            throw new UserException("Error " + e);
        }
    }

    @Override
    public User getUserByUsername(String username)
    {
        Connection connection = DBConnection.get().getConnection();
        User user = null;

        try
        {
            PreparedStatement stmt = connection.prepareStatement("SELECT * FROM users WHERE username = ?");
            stmt.setString(1, username);
            ResultSet rs = stmt.executeQuery();
            if (rs.next())
            {
                int id = rs.getInt("id");
                boolean is_librarian = rs.getInt("is_librarian") == 1;
                user = new User(id, rs.getString("username"), rs.getString("password"), is_librarian, getBorrowedBooks(id));
            }
            stmt.close();
        }
        catch (SQLException e)
        {
            throw new RepositoryException("Error " + e);
        }

        return user;
    }
}
